### PR TITLE
refactor: add mapping for ConnectorError in payouts flow

### PR DESCRIPTION
**Pull request by kashif-m**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Issue - #2605 

### Additional Changes

- [ ] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Steps to get your environment ready
- Head to adyen_uk / wise postman collection
- Head to QuickStart
- Create a merchant
- Create an API Key
- Create a payout connector [adyen]
- Create a payout connector [wise]
- All set!

##### Test case - Variation Cases / Scenario#10 / ACH Payout
Screenshot 2023-10-17 165156

##### Test case - Variation Cases / Scenario#10 / Bacs Payout
Screenshot 2023-10-17 165210

##### Test case - Variation Cases / Scenario#1 / Payouts creation
Screenshot 2023-10-17 165832

## NOTE
One of the test case assertions for payout API is disabled for Adyen. This is due to insufficient roles attached on the credentials which blocks any payout creation / submission at Adyen's end.
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [x] I added unit tests for my changes where possible
- [x] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: postman/collection-dir/adyen_uk/Flow Testcases/Happy Cases/.meta.json
Comment (pixincreate) [inline postman/collection-dir/adyen_uk/Flow Testcases/Happy Cases/.meta.json:1]: Can you fix the order?

File: postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json
Comment (pixincreate) [inline postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json:1]: Can they both (payment and payout) co-exist?

File: postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Payout Connector - Create/request.json [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
...9ZVzRnUVhCd2JHVWdhWE1nY0hKdmFHbGlhWFJsWkM0ZwpVbVZtWlhJZ2RHOGdkR2hsSUdGd2NHeHBZMkZpYkdVZ2MzUmhibVJoY21RZ2RHVnliWE1nWVc1a0lHTnZibVJwCmRHbHZibk1nYjJZZ2RYTmxMQ0JqWlhKMGFXWnBZMkYwWlNCd2IyeHBZM2tnWVc1a0lHTmxjblJwWm1sallYUnAKYjI0Z2NISmhZM1JwWTJVZ2MzUmhkR1Z0Wlc1MGN5NHdOd1lJS3dZQkJRVUhBZ0VXSzJoMGRIQnpPaTh2ZDNkMwpMbUZ3Y0d4bExtTnZiUzlqWlhKMGFXWnBZMkYwWldGMWRHaHZjbWwwZVM4d0V3WURWUjBsQkF3d0NnWUlLd1lCCkJRVUhBd0l3SFFZRFZSME9CQllFRk5RSysxcUNHbDRTQ1p6SzFSUmpnb05nM0hmdk1BNEdBMVVkRHdFQi93UUUKQXdJSGdEQlBCZ2txaGtpRzkyTmtCaUFFUWd4QVFVUkNRemxDTmtGRE5USkVRems0TnpCRk5qYzJNVFpFUkRJdwpPVUkwTWtReE1UVXlSVVpFTURVeFFVRXhRekV6T0ROR00wUkROa1V5TkVNelFqRkVSVEFQQmdrcWhraUc5Mk5rCkJpNEVBZ1VBTUEwR0NTcUdTSWIzRFFFQkN3VUFBNElCQVFBSFR6NTU2RUs5VVp6R0RVd2cvcmFibmYrUXFSYkgKcllVS0ZNcWQwUDhFTHZGMmYrTzN0ZXlDWHNBckF4TmVMY2hRSGVTNUFJOHd2azdMQ0xLUmJCdWJQQy9NVmtBKwpCZ3h5STg2ejJOVUNDWml4QVM1d2JFQWJYOStVMFp2RHp5Y01BbUNrdVVHZjNwWXR5TDNDaEplSGRvOEwwdmdvCnJQWElUSzc4ZjQzenNzYjBTNE5xbTE0eS9LNCs1ZkUzcFUxdEJqME5tUmdKUVJLRnB6MENpV2RPd1BRTk5BYUMKYUNNU2NiYXdwUTBjWEhaZDJWVjNtem4xdUlITEVtaU5GTWVxeEprbjZhUXFFQnNndDUzaUFxcmZMNjEzWStScAppd0tENXVmeU0wYzBweTYyZmkvWEwwS2c4ajEwWU1VdWJpd2dHajAzZThQWTB6bWUvcGZCZ3p6VQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==",
```
Comment (pixincreate) [inline postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Payout Connector - Create/request.json:306]: Remove the certificates. Everywhere.

File: postman/collection-dir/wise/.info.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    "schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json",
+    "_exporter_id": "9906252",
+    "_collection_link": "https://galactic-capsule-229427.postman.co/workspace/My-Workspace~2b563e0d-bad3-420f-8c0b-0fd5b278a4fe/collection/9906252-b5107328-6e3c-4ef0-b575-4072bc64462a?action=share&source=collection_link&creator=9906252"
```
Comment (pixincreate) [inline postman/collection-dir/wise/.info.json:8]: Please remove collection ID.

File: postman/collection-dir/wise/.variable.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
+{
+  "variable": [
+    {
+      "key": "baseUrl",
+      "value": "https://integ.hyperswitch.io",
```
Comment (pixincreate) [inline postman/collection-dir/wise/.variable.json:5]: Even this is not needed since it is updated in runtime

File: postman/collection-dir/wise/Flow Testcases/QuickStart/Payout Connector - Create/request.json [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
...9ZVzRnUVhCd2JHVWdhWE1nY0hKdmFHbGlhWFJsWkM0ZwpVbVZtWlhJZ2RHOGdkR2hsSUdGd2NHeHBZMkZpYkdVZ2MzUmhibVJoY21RZ2RHVnliWE1nWVc1a0lHTnZibVJwCmRHbHZibk1nYjJZZ2RYTmxMQ0JqWlhKMGFXWnBZMkYwWlNCd2IyeHBZM2tnWVc1a0lHTmxjblJwWm1sallYUnAKYjI0Z2NISmhZM1JwWTJVZ2MzUmhkR1Z0Wlc1MGN5NHdOd1lJS3dZQkJRVUhBZ0VXSzJoMGRIQnpPaTh2ZDNkMwpMbUZ3Y0d4bExtTnZiUzlqWlhKMGFXWnBZMkYwWldGMWRHaHZjbWwwZVM4d0V3WURWUjBsQkF3d0NnWUlLd1lCCkJRVUhBd0l3SFFZRFZSME9CQllFRk5RSysxcUNHbDRTQ1p6SzFSUmpnb05nM0hmdk1BNEdBMVVkRHdFQi93UUUKQXdJSGdEQlBCZ2txaGtpRzkyTmtCaUFFUWd4QVFVUkNRemxDTmtGRE5USkVRems0TnpCRk5qYzJNVFpFUkRJdwpPVUkwTWtReE1UVXlSVVpFTURVeFFVRXhRekV6T0ROR00wUkROa1V5TkVNelFqRkVSVEFQQmdrcWhraUc5Mk5rCkJpNEVBZ1VBTUEwR0NTcUdTSWIzRFFFQkN3VUFBNElCQVFBSFR6NTU2RUs5VVp6R0RVd2cvcmFibmYrUXFSYkgKcllVS0ZNcWQwUDhFTHZGMmYrTzN0ZXlDWHNBckF4TmVMY2hRSGVTNUFJOHd2azdMQ0xLUmJCdWJQQy9NVmtBKwpCZ3h5STg2ejJOVUNDWml4QVM1d2JFQWJYOStVMFp2RHp5Y01BbUNrdVVHZjNwWXR5TDNDaEplSGRvOEwwdmdvCnJQWElUSzc4ZjQzenNzYjBTNE5xbTE0eS9LNCs1ZkUzcFUxdEJqME5tUmdKUVJLRnB6MENpV2RPd1BRTk5BYUMKYUNNU2NiYXdwUTBjWEhaZDJWVjNtem4xdUlITEVtaU5GTWVxeEprbjZhUXFFQnNndDUzaUFxcmZMNjEzWStScAppd0tENXVmeU0wYzBweTYyZmkvWEwwS2c4ajEwWU1VdWJpd2dHajAzZThQWTB6bWUvcGZCZ3p6VQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==",
```
Comment (pixincreate) [inline postman/collection-dir/wise/Flow Testcases/QuickStart/Payout Connector - Create/request.json:305]: Remove the certificates

File: postman/collection-json/wise.postman_collection.json
Comment (pixincreate) [inline postman/collection-json/wise.postman_collection.json:1]: Having directory is enough since github actions will however generate it again even if you put it here

**Review by pixincreate — CHANGES_REQUESTED** [review]

**Comment by pixincreate** [comment]
Kindly update this section with relevant changes that you did:
image

If you tested the code through collection, and the collection is added in code, kindly add screenshots of them passing and the command you used to run. A brief description of how this can be verified on an envrionment like Sandbox would be enough.

File: postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json
Comment (kashif-m) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json:1]: Yes, same connectors can be co-exist. This is controlled by connector_type field while creating the connector

Comment (pixincreate) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json:1]: > This is controlled by connector_type field while creating the connector

Wouldn't this would cause conflicts in adyen like collections that support both payment as well as payout since we're passing in only Connector auth that related to payment and not payout (Yes, I haven't checked whether the auth is same for both the scenarious)?

Comment (kashif-m) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json:1]: Same connector can be used for both payments and payouts. Although, the auth for both of these is different. Payments uses BodyKey, whereas payouts uses SignatureKey. 

Moreover, this wouldn't cause any conflicts since connector_type for payouts flow must always be `payout_processor`, so we filter out all the unsupported connectors that way

Comment (pixincreate) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json:1]: So, we've moved the Adyen collection to use `Signaturekey` for postman collection long back. That should not be an issue. However, I'm afraid regarding adyen postman tests running with 0 failures. Can you please run the collection using `rustman` once by referring to this [documentation](https://github.com/juspay/hyperswitch/blob/main/crates/test_utils/README.md)? 
TL;Dr: Start your server from this branch, open another terminal tab and run the command mentioned in the doc.

Comment (kashif-m) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json:1]: Sure, will test this.

Also, can we add any such test cases as a part of PR merge checks in Github Actions?

Comment (pixincreate) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/.meta.json:1]: Currently, we're running only Stripe's collection that also checks for all the hyperswitch supported features (yes, we're a bit behind regarding that, but constantly being updated every time a new feature is added). So, if the feature that you added is supported by Stripe, then that will also run on every PR raised.

File: postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Merchant Account - Create/request.json [hunk 5]
```diff
@@ -79,6 +83,14 @@
       "metadata": {
         "city": "NY",
         "unit": "245"
+      },
+      "routing_algorithm": {
```
Comment (pixincreate) [inline postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Merchant Account - Create/request.json:87]: Adding `routing_algorithm` will cause the collection to fail. Kindly remove it. Instead, add it in `Payment Create` step.

**Review by pixincreate — CHANGES_REQUESTED** [review]

File: postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Merchant Account - Create/request.json (hunk 5 above)
Comment (kashif-m) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Merchant Account - Create/request.json:87]: Why would the collection fail? I have tested this locally, adding this here means there's no need to pass this individually in txn requests
image

Comment (pixincreate) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Merchant Account - Create/request.json:87]: You can refer this PR for that: https://github.com/juspay/hyperswitch/pull/2032

Comment (pixincreate) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Merchant Account - Create/request.json:87]: I ran the collection locally, 14 assertions failed. 4 failures is related to manual capture being not allowed from the connector, and the rest 10 failures is related to refunds not being processed (They're going to `pending` state instead of `succeeded`. Same is happening with `Stripe` as well which can be seen below in the CI checks).

Comment (kashif-m) [reply postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Merchant Account - Create/request.json]: I have moved routing_algorithm to payment and payout create requests.

image

File: crates/diesel_models/src/payout_attempt.rs [hunk 6]
```diff
@@ -26,7 +26,7 @@ pub struct PayoutAttempt {
     pub created_at: PrimitiveDateTime,
     #[serde(with = "common_utils::custom_serde::iso8601")]
     pub last_modified_at: PrimitiveDateTime,
-    pub profile_id: Option<String>,
+    pub profile_id: String,
```
Comment (pixincreate) [inline crates/diesel_models/src/payout_attempt.rs:29]: backfilling of `profile_id`s done? @Narayanbhat166

File: postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Payouts - Create/event.test.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+});
+
+// Validate if status is successful
+// if (jsonData?.status) {
+//   pm.test("[POST]::/payouts/create - Content check if value for 'status' matches 'success'",
+//     function () {
+//       pm.expect(jsonData.status).to.eql("success");
+//     },
+//   );
+// }
```
Comment (pixincreate) [inline postman/collection-dir/adyen_uk/Flow Testcases/QuickStart/Payouts - Create/event.test.js:28]: Marking this, which needs to be uncommented once the things are sorted out.  
CC: @bernard-eugine

**Review by pixincreate — APPROVED** [review]

File: migrations/2023-10-17-084554_alter_payout_identifier_length/up.sql [hunk 8]
```diff
@@ -0,0 +1,24 @@
+-- Payout Attempt table
+ALTER TABLE
```
Comment (jagan-jaya) [inline migrations/2023-10-17-084554_alter_payout_identifier_length/up.sql:2]: Alter table will result in DB downtime. So requests might fail during the migration. 
Please check with @venkatesh-juspay on how do we go with this. 
Worst case solution would be creating new column with the new size. New column is not a problem as it won't cause downtime.

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: migrations/2023-10-17-084554_alter_payout_identifier_length/up.sql (hunk 8 above)
Comment (kashif-m) [reply migrations/2023-10-17-084554_alter_payout_identifier_length/up.sql]: Removed this change.

`payout_attempt_id` will be created using payout_id and attempt_count now.

image

**Review by jagan-jaya — APPROVED** [review]

**Review by pixincreate — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by pixincreate — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]
